### PR TITLE
[prometheus] aggregating proxy

### PR DESCRIPTION
**Pull request by pashcovich**
## Description

Aggregating proxy for HA Prometheuses based on  promxy and mimir query-fronted 
## Why do we need it, and what problem does it solve?

Close https://github.com/deckhouse/deckhouse/issues/6934
## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: prometheus
type: feature 
summary:  Aggregating proxy for Prometheus in HA mode.
impact: Now there is one Grafana datasource to see combined metrics from the Main and the Longterm Prometheus.
impact_level: default
```

### CODE REVIEW

File: modules/300-prometheus/images/mimir/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ARG SOURCE_REPO
+
+ENV GOPROXY=${GOPROXY} \
+    SOURCE_REPO=${SOURCE_REPO} \
+    CGO_ENABLED=0 \
+    GOOS=linux \
+    GOARCH=amd64
+
+RUN apk add --no-cache make git
+RUN git clone --depth 1 --branch mimir-2.10.4 ${SOURCE_REPO}/grafana/mimir.git /mimir
```
Comment (nabokihms) [inline modules/300-prometheus/images/mimir/Dockerfile:16]: Could you please add new entries to the `oss.yaml` file in the module directory?

File: modules/300-prometheus/templates/grafana/secret-datasources-list.yaml [hunk 2]
```diff
@@ -46,6 +47,13 @@ datasources:
   isDefault: true
   {{- include "authorization_config" (list . (include "main_scrape_interval" . )) | nindent 2 }}
 
+- name: promxy
```
[the remark this answers is not in the document]
Comment (nabokihms) [reply modules/300-prometheus/templates/grafana/secret-datasources-list.yaml:50]: By the way, is there any hesitation to point the metrics adapter to the promxy?

File: modules/300-prometheus/templates/promxy/servicemonitor.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+---
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: promxy-module
```
Comment (nabokihms) [inline modules/300-prometheus/templates/promxy/servicemonitor.yaml:5]:
```suggestion
  name: promxy
```

File: modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            path: /ready
+            port: 8443
+            scheme: HTTPS
+        env:
+        - name: PROMETHEUS_TOKEN
+          valueFrom:
+            secretKeyRef:
+              name: prometheus-token
+              key: token
+              optional: false
```
Comment (nabokihms) [inline modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml:90]: How mimir-frontend and promxy use the variable? Didn't find it anywhere.

File: modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                resourceAttributes:
+                  namespace: d8-monitoring
+                  apiGroup: apps
+                  apiVersion: v1
+                  resource: deployments
+                  subresource: prometheus-metrics
+                  name: mimir-query-frontend
+        resources:
+          requests:
+            {{- include "helm_lib_module_ephemeral_storage_only_logs" . | nindent 12 }}
```
Comment (nabokihms) [inline modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml:139]: If tmp dir is mounted into an empty dir, specifying more ephemeral storage requests is required.

File: modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      volumes:
+      - name: mimir-config
+        configMap:
+          name: mimir-config
+      - name: tmp
+        emptyDir: {}
+      - name: kube-rbac-proxy-ca
+        configMap:
+          defaultMode: 420
+          name: kube-rbac-proxy-ca.crt
```
Comment (nabokihms) [inline modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml:155]: Is this mount required? Didn't find where it is used. The same question goes to promxy.

File: modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml (hunk 4 above)
Comment (nabokihms) [reply modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml:90]: The same question goes to promxy.

File: modules/300-prometheus/docs/README.md [hunk 7]
```diff
@@ -73,13 +73,20 @@ Deckhouse supports sending alerts using `Alertmanager`:
 
 <img src="../../images/300-prometheus/prometheus_monitoring.svg">
 
+##### WIP: new scheme for single datasource (testing_now)
+
+<img src="../../images/300-prometheus/prometheus_monitoring_new.svg">
```
Comment (nabokihms) [inline modules/300-prometheus/docs/README.md:78]: Let's make it official an remove trickster.

File: docs/documentation/images/300-prometheus/prometheus_monitoring_new.svg
Comment (nabokihms) [inline docs/documentation/images/300-prometheus/prometheus_monitoring_new.svg:1]: NIT: Can we draw memcache using a different color?

**Review by nabokihms — CHANGES_REQUESTED** [review]

File: modules/300-prometheus/docs/README_RU.md [hunk 8]
```diff
@@ -74,17 +74,23 @@ Deckhouse поддерживает отправку алертов с помощ
...
+| Компонент                   | Описание                                                                                                                                                                                                                                                                                        |
+|-----------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| **prometheus-main**         | Основной Prometheus, который выполняет scrape каждые 30 секунд (с помощью параметра `scrapeInterval` можно изменить это значение). Именно он обрабатывает все правила, отправляет алерты и является основным источником данных.                                                                 |
```
Comment (MariaMakeeva) [inline modules/300-prometheus/docs/README_RU.md:85]:
```suggestion
| **prometheus-main**         | Основной Prometheus, который выполняет scrape каждые 30 секунд (с помощью параметра `scrapeInterval` можно изменить это значение). Он обрабатывает все правила, отправляет предупреждения и является основным источником данных.                                                                 |
```

File: modules/300-prometheus/docs/README_RU.md [hunk 9]
```diff
@@ -74,17 +74,23 @@ Deckhouse поддерживает отправку алертов с помощ
...
+| **aggregating-proxy**       | Аггрегирующий и кеширующий прокси, снижающий нагрузку на Prometheus и объединяющий main и longterm в один источник.                                                                                                                                                                             |
+| **memcached**               | Сервис кэширования данных в оперативной памяти.                                                                                                                                                                                                                                                 |
+| **grafana**                 | Управляемая платформа визуализации данных. Включает подготовленные dashboard'ы для всех модулей Deckhouse и некоторых популярных приложений. Grafana умеет работать в режиме высокой доступности, не хранит состояние и настраивается с помощью CRD.                                            |
```
Comment (MariaMakeeva) [inline modules/300-prometheus/docs/README_RU.md:90]:
```suggestion
| **grafana**                 | Управляемая платформа визуализации и мониторинга данных. Включает предварительно разработанные dashboard'ы для всех модулей Deckhouse и нескольких распространенных приложений. Grafana работает в режиме высокой доступности, не хранит состояние и настраивается с помощью CRD.                                            |
```

File: docs/documentation/images/300-prometheus/prometheus_monitoring_new.svg
Comment (pashcovich) [reply docs/documentation/images/300-prometheus/prometheus_monitoring_new.svg:1]: redrawn in blue

File: modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+  updatePolicy:
+    updateMode: "Auto"
+  resourcePolicy:
+    containerPolicies:
+    - containerName: mimir-query-frontend
+      minAllowed:
+        {{- include "mimir-query-frontend_resources" . | nindent 8 }}
+      maxAllowed:
+        cpu: 50m
+        memory: 50Mi
```
[the remark this answers is not in the document]
Comment (pashcovich) [reply modules/300-prometheus/templates/mimir-query-frontend/deployment.yaml]: slightly increased, but the values are taken from the test environment

File: modules/300-prometheus/templates/grafana/secret-datasources-list.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pashcovich) [reply modules/300-prometheus/templates/grafana/secret-datasources-list.yaml]: renamed to main and main renamed to main-old
and yes, i think it`s better to test before pointing pma to new scheme

File: modules/300-prometheus/docs/README.md (hunk 7 above)
Comment (pashcovich) [reply modules/300-prometheus/docs/README.md]: in next iteration

File: modules/300-prometheus/docs/README_RU.md (hunk 9 above)
Comment (pashcovich) [reply modules/300-prometheus/docs/README_RU.md:90]: I do not agree with these changes

Comment (pashcovich) [reply modules/300-prometheus/docs/README_RU.md:90]: the word `работает` significantly changes the meaning of a sentence and can be misleading

File: modules/300-prometheus/docs/README_RU.md (hunk 8 above)
Comment (pashcovich) [reply modules/300-prometheus/docs/README_RU.md:85]: Prometheus sends [alerts](https://prometheus.io/docs/practices/alerting/), not notifications, warnings or anything else. So "предупреждения", I think, is the wrong word here.

File: modules/300-prometheus/docs/README_RU.md [hunk 11]
```diff
@@ -74,17 +74,23 @@ Deckhouse поддерживает отправку алертов с помощ
 
 <img src="../../images/300-prometheus/prometheus_monitoring.svg">
 
+WIP:  новая схема для единого  датасорса
```
Comment (z9r5) [inline modules/300-prometheus/docs/README_RU.md:77]: @pashcovich Can we remove the WIP here?

Comment (pashcovich) [reply modules/300-prometheus/docs/README_RU.md:77]: I think only if we point the PMA to a new scheme

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

File: modules/300-prometheus/docs/README_RU.md [hunk 12]
```diff
@@ -74,17 +74,23 @@ Deckhouse поддерживает отправку алертов с помощ
...
+
+<img src="../../images/300-prometheus/prometheus_monitoring_new.svg">
+
```
Comment (z9r5) [inline modules/300-prometheus/docs/README_RU.md:80]:
```suggestion
Схема взаимодействия модулей:

![Схема взаимодействия модулей]](../../images/300-prometheus/prometheus_monitoring.svg)

Схема взаимодействия модулей с использованием одного агрегированного источника данных (WIP):  

![Схема взаимодействия модулей с использованием одного агрегированного источника данных (WIP)](../../images/300-prometheus/prometheus_monitoring_new.svg)

```

File: modules/300-prometheus/docs/README.md [hunk 13]
```diff
@@ -73,13 +73,19 @@ Deckhouse supports sending alerts using `Alertmanager`:
 
 <img src="../../images/300-prometheus/prometheus_monitoring.svg">
 
+WIP: new scheme for single datasource (testing_now)
+
+<img src="../../images/300-prometheus/prometheus_monitoring_new.svg">
+
```
Comment (z9r5) [inline modules/300-prometheus/docs/README.md:79]:
```suggestion
Module interaction scheme:

![Module interaction scheme](../../images/300-prometheus/prometheus_monitoring.svg)

The scheme of interaction of modules using a single aggregated data source (WIP):

![The scheme of interaction of modules using a single aggregated data source (WIP)](../../images/300-prometheus/prometheus_monitoring_new.svg)

```

**Comment by z9r5** [comment]
Please [fix DCO tests](https://github.com/deckhouse/deckhouse/pull/6987/checks?check_run_id=20889851820) @pashcovich

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

File: modules/300-prometheus/docs/README_RU.md [hunk 14]
```diff
@@ -72,19 +72,27 @@ Deckhouse поддерживает отправку алертов с помощ
 
 ## Включенные модули
 
-<img src="../../images/300-prometheus/prometheus_monitoring.svg">
+Схема взаимодействия модулей:
+
+![Схема взаимодействия модулей]](../../images/300-prometheus/prometheus_monitoring.svg)
```
Comment (z9r5) [inline modules/300-prometheus/docs/README_RU.md:77]:
```suggestion
![Схема взаимодействия модулей](../../images/300-prometheus/prometheus_monitoring.svg)
```

**Review by z9r5 — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

**Review by z9r5 — APPROVED** [review]

File: modules/300-prometheus/oss.yaml [hunk 15]
```diff
@@ -16,4 +16,18 @@
   logo: https://raw.githubusercontent.com/trickstercache/trickster/master/docs/images/logos/trickster-logo.svg
   licence: Apache License 2.0
 
+- name: Mimir
+  link: https://github.com/trickstercache/trickster
```
Comment (nabokihms) [inline modules/300-prometheus/oss.yaml:20]: The link leads to the trickster website.

File: modules/300-prometheus/oss.yaml [hunk 16]
```diff
@@ -16,4 +16,18 @@
...
+  description: A HTTP reverse proxy/cache for http applications and a dashboard query accelerator for time series databases.
+  logo: https://grafana.com/static/img/logos/logo-mimir.svg
+  licence: AGPL-3.0 license
+
+- name: Promxy
+  link: https://github.com/jacksontj/promxy
+  description: An aggregating proxy to enable HA prometheus
```
Comment (nabokihms) [inline modules/300-prometheus/oss.yaml:27]: logo is missing

File: modules/300-prometheus/oss.yaml [hunk 17]
```diff
@@ -16,4 +16,18 @@
...
+  licence: MIT License
 
+- name: Memcached
+  link: https://github.com/memcached/memcached
```
Comment (nabokihms) [inline modules/300-prometheus/oss.yaml:31]: logo is missing

File: modules/300-prometheus/oss.yaml (hunk 16 above)
Comment (nabokihms) [inline modules/300-prometheus/oss.yaml:27]: The description does not describe what promxy actually does. There is a better version on its website.

> Promxy is a prometheus proxy that makes many shards of prometheus appear as a single API endpoint to the user. This significantly simplifies operations and use of prometheus at scale (when you have more than one prometheus host).

File: modules/300-prometheus/docs/README_RU.md [hunk 18]
```diff
@@ -72,19 +72,27 @@ Deckhouse поддерживает отправку алертов с помощ
...
 ## Включенные модули
 
-<img src="../../images/300-prometheus/prometheus_monitoring.svg">
+Схема взаимодействия модулей:
+
+![Схема взаимодействия модулей](../../images/300-prometheus/prometheus_monitoring.svg)
+
+Схема взаимодействия модулей с использованием одного агрегированного источника данных (WIP):  
+
+![Схема взаимодействия модулей с использованием одного агрегированного источника данных (WIP)](../../images/300-prometheus/prometheus_monitoring_new.svg)
```
Comment (nabokihms) [inline modules/300-prometheus/docs/README_RU.md:81]:
```suggestion
<img src="../../images/300-prometheus/prometheus_monitoring_new.svg">
```

File: modules/300-prometheus/docs/README.md [hunk 19]
```diff
@@ -71,7 +71,13 @@ Deckhouse supports sending alerts using `Alertmanager`:
...
 ## Included modules
 
-<img src="../../images/300-prometheus/prometheus_monitoring.svg">
+Module interaction scheme:
+
+![Module interaction scheme](../../images/300-prometheus/prometheus_monitoring.svg)
+
+The scheme of interaction of modules using a single aggregated data source (WIP):
+
+![The scheme of interaction of modules using a single aggregated data source (WIP)](../../images/300-prometheus/prometheus_monitoring_new.svg)
```
Comment (nabokihms) [inline modules/300-prometheus/docs/README.md:80]:
```suggestion
<img src="../../images/300-prometheus/prometheus_monitoring_new.svg">
```

File: modules/300-prometheus/templates/aggregating-proxy/servicemonitor-mimir.yaml [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+    honorLabels: true
+    relabelings:
+    - targetLabel: tier
+      replacement: cluster
+  selector:
+    matchLabels:
+      app: aggregating-proxy
+  namespaceSelector:
+    matchNames:
+    - d8-monitoring
```
Comment (nabokihms) [inline modules/300-prometheus/templates/aggregating-proxy/servicemonitor-mimir.yaml:26]: Probably, it is best to use a single service monitor with two endpoints then having two separated for the same set of pods. 

There is an example https://github.com/deckhouse/deckhouse/blob/main/modules/340-monitoring-kubernetes/templates/kubelet/service-monitor.yaml

**Review by nabokihms — CHANGES_REQUESTED** [review]

File: modules/300-prometheus/oss.yaml (hunk 16 above)
Comment (pashcovich) [reply modules/300-prometheus/oss.yaml]: added

File: modules/300-prometheus/oss.yaml (hunk 17 above)
Comment (pashcovich) [reply modules/300-prometheus/oss.yaml:32]: added

File: modules/300-prometheus/oss.yaml (hunk 16 above)
Comment (pashcovich) [reply modules/300-prometheus/oss.yaml]: changed

File: modules/300-prometheus/templates/aggregating-proxy/servicemonitor-mimir.yaml (hunk 20 above)
Comment (pashcovich) [reply modules/300-prometheus/templates/aggregating-proxy/servicemonitor-mimir.yaml]: thank you for example, changed

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
Please rebase it

**Pull request closed without merging**